ledger/ledgercore/statedelta.go

```diff
@@ -149,3 +152,31 @@ func (ad *AccountDeltas) upsert(br basics.BalanceRecord) {
 	}
 	ad.acctsCache[addr] = last
 }
+
+func (sd *StateDelta) Compress() {
+	sd.Accts.accts = sd.Accts.accts[:len(sd.Accts.accts)]
+
+	if len(sd.Accts.acctsCache) < sd.initialTransactionsCount / 2 {
```

Comment:
The idea is correct, but you're not getting full-value for your efforts. What you want is to make this condition evaluate the memory cost gains of the reallocation before you do it.
in the case of the map, you want to do something like this:
```
initial_cost = sd.initialTransactionsCount * sizeofSuchAMapPerElement
current_cost = len(sd.Accts.acctsCache) * sizeofSuchAMapPerElement
if  initial_cost > current_cost && initial_cost-current_cost > minGainsTreshold {
   .. re-allocate..
}
```

create memory benchmark to measure the sizeofSuchAMapPerElement, and figure out a reasonable value for minGainsTreshold.

Note that when performing this trick on an array, you need to use the array's cap to figure out its underlying array size. The len would only give you the slice length.